### PR TITLE
Loopback mode kcontrol support

### PR DESCRIPTION
**Pull request by xiulipan**
Add topology parser for loopback mode kcontrol and switch kcontrol callback function for loopback mode kcontrol switch.
work with:
https://github.com/thesofproject/sof/pull/26
https://github.com/thesofproject/soft/pull/13

### CODE REVIEW

File: sound/soc/sof/debug.c [hunk 1]
```diff
@@ -126,6 +127,19 @@ int snd_sof_dbg_init(struct snd_sof_dev *sdev)
 				map->name);
 	}
 
+#ifdef DEBUG
+	/* enable debug mode for DEBUG build */
+	sdev->debug_mode = 1;
+#else
+	sdev->debug_mode = 0;
```
Comment (lgirdwood) [inline sound/soc/sof/debug.c:134]: there might not be need for this if we bzero sdev on allocation ?

File: sound/soc/sof/control.c [hunk 2]
```diff
@@ -195,3 +195,99 @@ int snd_sof_bytes_put(struct snd_kcontrol *kcontrol,
...
+			     struct snd_ctl_elem_value *ucontrol)
+{
+	struct snd_sof_control *scontrol =
+		(struct snd_sof_control *)kcontrol->private_value;
+	struct snd_sof_dev *sdev = scontrol->sdev;
+
+	/* can only switch when debug_mode enable */
+	if (!sdev->debug_mode) {
+		dev_err(sdev->dev,
+			"debug switch only enabled with debugfs debug_mode nonzero!\n");
```
Comment (lgirdwood) [inline sound/soc/sof/control.c:278]: can this be more meaning full, like "error: kcontrol %s can only be changed in driver debug mode\n");

File: sound/soc/sof/topology.c [hunk 3]
```diff
@@ -804,6 +817,70 @@ static int sof_connect_dai_widget(struct snd_soc_component *scomp,
 	return 0;
 }
 
+/* LBM kcontrol, used to control loopback mode on SSP */
+static int sof_dai_add_lbm_kcontrol(struct snd_soc_component *scomp,
```
Comment (lgirdwood) [inline sound/soc/sof/topology.c:821]: This is completely generic code except for the strings and provate data. These should be passed in as args.

File: sound/soc/sof/topology.c [hunk 4]
```diff
@@ -804,6 +817,70 @@ static int sof_connect_dai_widget(struct snd_soc_component *scomp,
...
+				    u32 comp_id, u32 index)
+{
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
+	struct snd_sof_control *scontrol;
+	struct sof_ipc_ctrl_data *cdata;
+	struct snd_kcontrol *kcontrol;
+	struct snd_kcontrol_new *kc;
+	char kcname[50];
```
Comment (lgirdwood) [inline sound/soc/sof/topology.c:829]: please align the size with ALSA  asounds.h/control.h

File: sound/soc/sof/debug.c (hunk 1 above)
Comment (xiulipan) [reply sound/soc/sof/debug.c:134]: If we bzero sdev, then it is not needed.
I will check that.
Thanks for reminding.

File: sound/soc/sof/control.c (hunk 2 above)
Comment (xiulipan) [reply sound/soc/sof/control.c:278]: Will change to that.

File: sound/soc/sof/topology.c (hunk 3 above)
Comment (xiulipan) [reply sound/soc/sof/topology.c:821]: Thought about that.
Will do that.

File: sound/soc/sof/topology.c (hunk 4 above)
Comment (xiulipan) [reply sound/soc/sof/topology.c:829]: Will check that. Thanks

**Comment by plbossart** [comment]
closing, no update or follow-up

**Pull request closed without merging**